runtime-sdk/modules/evm/src/precompile/erc20.rs

```diff
@@ -0,0 +1,552 @@
+where
+    T: AccountToken,
+    F: FnOnce(Option<u128>) -> Result<u128, Error>,
+{
+    let owner = owner.as_bytes();
+    let spender = spender.as_bytes();
+    CurrentState::with_store(|store| {
+        let store = storage::PrefixStore::new(store, &MODULE_NAME);
+        let allowances = storage::PrefixStore::new(store, &STATE_ALLOWANCES);
+        let denomination = T::denomination().to_string();
+        let allowances = storage::PrefixStore::new(allowances, &denomination);
+        let mut owner = storage::TypedStore::new(storage::PrefixStore::new(allowances, &owner));
+        let old_allowance = owner.get(spender);
+        let new_allowance = f(old_allowance)?;
+        if new_allowance == 0 {
+            owner.remove(spender);
+        } else {
+            owner.insert(spender, new_allowance);
+        }
+        Ok(old_allowance.unwrap_or_default())
+    })
+}
+
+impl<T: AccountToken> Erc20Token for T {
+    const GAS_COSTS: TokenOperationCosts = T::GAS_COSTS;
+
+    const NAME: &str = T::NAME;
+    const SYMBOL: &str = T::SYMBOL;
+    const DECIMALS: u8 = T::DECIMALS;
+
+    fn address() -> H160 {
+        Self::ADDRESS
+    }
+
+    fn total_supply() -> Result<u128, Error> {
+        Ok(T::Accounts::get_total_supplies()?
+            .get(&T::denomination())
+            .copied()
+            .unwrap_or_default())
+    }
+
+    fn balance_of(account: &H160) -> Result<u128, Error> {
+        let account = address::Address::from_eth(account.as_bytes());
+        Ok(T::Accounts::get_balance(account, T::denomination())?)
+    }
+
+    fn transfer(sender: &H160, recipient: &H160, amount: u128) -> Result<bool, Error> {
+        if sender.is_zero() {
+            return Err(Error::ERC20InvalidSender(*sender));
+        }
+        if recipient.is_zero() {
+            return Err(Error::ERC20InvalidReceiver(*recipient));
+        }
+        let amount = token::BaseUnits::new(amount, T::denomination());
+        let sender_address = address::Address::from_eth(sender.as_bytes());
+        let recipient_address = address::Address::from_eth(recipient.as_bytes());
+        T::Accounts::transfer(sender_address, recipient_address, &amount)
+            .map(|_| true)
+            .map_err(|e| Error::annotate_accounts_error::<T>(e, sender_address, *sender, &amount))
+    }
+
+    fn transfer_from(
+        owner: &H160,
+        spender: &H160,
+        recipient: &H160,
+        amount: u128,
+    ) -> Result<bool, Error> {
+        if owner.is_zero() {
+            return Err(Error::ERC20InvalidApprover(*owner));
+        }
+        if spender.is_zero() {
+            return Err(Error::ERC20InvalidSpender(*spender));
+        }
+        if recipient.is_zero() {
+            return Err(Error::ERC20InvalidReceiver(*recipient));
+        }
+        with_allowance::<Self, _>(owner, spender, |allowance| {
+            let allowance = allowance.unwrap_or_default();
+            if allowance == u128::MAX {
```

Comment:
Feel free to remove this, as you pointed out in https://github.com/oasisprotocol/oasis-sdk/pull/2144#discussion_r1971391385 the amounts in EMV are U256, so maybe it's better we don't do this and it also simplifies the code a bit.